python/lsst/resources/_resourceHandles/_s3ResourceHandle.py

```diff
@@ -0,0 +1,260 @@
+        self._closed = CloseStatus.CLOSED
+
+    @property
+    def closed(self) -> bool:
+        return self._closed == CloseStatus.CLOSED
+
+    def fileno(self) -> int:
+        raise OSError("S3 object does not have a file number")
+
+    def flush(self) -> None:
+        # If the object is closed, not writeable, or rw flush should be skipped
+        # rw mode skips flush because the whole bytestream must be kept in
+        # the buffer for seeking reasons.
+        if self.closed or not self.writable() or "+" in self._mode:
+            return
+        # Disallow writes to seek to a position prior to the previous flush
+        # this allows multipart uploads to upload content as the stream is
+        # written to.
+        if (
+            (self.tell() - (self._last_flush_position or 0)) < 5242880
+            and not self._closed == CloseStatus.CLOSING
+            and not self._warned
+        ):
+            warnings.warn("S3 does not support flushing objects less than 5Mb, skipping")
+            self._warned = True
+            return
+        # nothing to write, don't create an empty upload
+        if self.tell() == 0:
+            return
+        with time_this(
+            self._log,
+            msg="Upload multipart %d to %s",
+            args=(
+                self._partNo,
+                self,
+            ),
+        ):
+            response = self._client.upload_part(
+                Body=self._buffer.getvalue(),
+                Bucket=self._bucket,
+                Key=self._key,
+                UploadId=self._multiPartUpload["UploadId"],
+                PartNumber=self._partNo,
+            )
+        self._parts.append({"PartNumber": self._partNo, "ETag": response["ETag"]})
+        self._partNo += 1
+        self._last_flush_position = self._buffer.tell() + (self._last_flush_position or 0)
+        self._buffer = BytesIO()
+
+    @property
+    def isatty(self) -> bool:
+        return False
+
+    def readable(self) -> bool:
+        return self._readable
+
+    def readline(self, size: int = -1) -> bytes:
+        raise OSError("S3 Does not support line by line reads")
+
+    def readlines(self, hint: int = -1) -> Iterable[bytes]:
+        self.seek(0)
+        return self.read().split(self._newline)
+
+    def seek(self, offset: int, whence: int = SEEK_SET) -> int:
+        if self.writable():
+            if self._last_flush_position is not None:
+                if whence == SEEK_SET:
+                    offset -= self._last_flush_position
+                    if offset < 0:
+                        raise OSError("S3 ResourceHandle can not seek prior to already flushed positions")
+                if whence == SEEK_CUR:
+                    if (self.tell() - self._last_flush_position) < 0:
+                        raise OSError("S3 ResourceHandle can not seek prior to already flushed positions")
```

Comment:
For multipart uploading (which is what flush does) is it possible to invalidate parts that have been uploaded before the multipart upload is finalized?

Reply:
I worked through this quite a lot to find different solutions and settled on this, because after you hand off to have things uploaded it becomes a whole song and dance to find out what actually has been written out by the client, what to do if things are in flight etc. We can add that support, but I would prefer to do it on another ticket.

Reply:
Ok. File a ticket.